### PR TITLE
8281455: Change JVM options with small ranges from 64 to 32 bits, for gc_globals.hpp

### PR DESCRIPTION
**Pull request by afshin-zafari**
The `uintx/intx` options whose ranges are small enough are changed to `uint/int`, otherwise gcc complains 
when `-Wconversion` is used in build.
Their uses in printf formats are changed accordingly.

### Tests
Locally hotspot:tier1 tested on linux-x64
mach5 tiers 1-4 on Linux and Windows 64bits platforms passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8281455](https://bugs.openjdk.org/browse/JDK-8281455): Change JVM options with small ranges from 64 to 32 bits, for gc_globals.hpp (**Enhancement** - P4)

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14259/head:pull/14259` \
`$ git checkout pull/14259`

Update a local copy of the PR: \
`$ git checkout pull/14259` \
`$ git pull https://git.openjdk.org/jdk.git pull/14259/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14259`

View PR using the GUI difftool: \
`$ git pr show -t 14259`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14259.diff">https://git.openjdk.org/jdk/pull/14259.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14259#issuecomment-1580237728)

### CODE REVIEW

**Comment by tschatzl** [comment]
Although it's a bit early to review, when I looked at this briefly, I believe also the `uintx` that have a `juint_max` upper range can be replaced with `uint` (and use the equivalent `UINT_MAX` as upper range, `j`-types should only be used with JNI code).

**Comment by afshin-zafari** [comment]
> Although it's a bit early to review, when I looked at this briefly, I believe also the `uintx` that have a `juint_max` upper range can be replaced with `uint` (and use the equivalent `UINT_MAX` as upper range, `j`-types should only be used with JNI code).

Done.

**Comment by offamitkumar** [comment]
Wouldn't that be good to keep the alignment for `\` symbol in `gc_global.hpp` file ?

File: src/hotspot/share/gc/parallel/psAdaptiveSizePolicy.cpp [hunk 1]
```diff
@@ -349,7 +349,7 @@ void PSAdaptiveSizePolicy::compute_eden_space_size(
     log_debug(gc, ergo)(
           "PSAdaptiveSizePolicy::compute_eden_space_size: gc time limit"
           " gc_cost: %f "
-          " GCTimeLimit: " UINTX_FORMAT,
+          " GCTimeLimit: " UINT32_FORMAT,
```
Comment (tschatzl) [inline src/hotspot/share/gc/parallel/psAdaptiveSizePolicy.cpp:352]: We use `%u` as format specifiers for uint.

File: src/hotspot/share/gc/shared/gcOverheadChecker.cpp [hunk 2]
```diff
@@ -91,11 +91,11 @@ void GCOverheadChecker::check_gc_overhead_limit(GCOverheadTester* time_overhead,
 
   if (UseGCOverheadLimit) {
     if (gc_overhead_limit_exceeded()) {
-      log_trace(gc, ergo)("GC is exceeding overhead limit of " UINTX_FORMAT "%%", GCTimeLimit);
+      log_trace(gc, ergo)("GC is exceeding overhead limit of %d%%", GCTimeLimit);
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/gcOverheadChecker.cpp:94]: Not sure why this place uses `%d` as format specifier compared to the `UINT32_FORMAT` the previous places did. Use `%u` ;)

File: src/hotspot/share/gc/shared/gc_globals.hpp [hunk 3]
```diff
@@ -166,7 +166,7 @@
           "A System.gc() request invokes a concurrent collection; "         \
           "(effective only when using concurrent collectors)")              \
                                                                             \
-  product(uintx, GCLockerEdenExpansionPercent, 5,                           \
+  product(uint, GCLockerEdenExpansionPercent, 5,                           \
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/gc_globals.hpp:169]: The `\` needs to be aligned back.

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 4]
```diff
@@ -431,13 +431,15 @@ const intx  min_intx  = (intx)1 << (sizeof(intx)*BitsPerByte-1);
...
 
+
 // Table of values:
 //      sizeof intx         4               8
 // min_intx             0x80000000      0x8000000000000000
 // max_intx             0x7FFFFFFF      0x7FFFFFFFFFFFFFFF
 // max_uintx            0xFFFFFFFF      0xFFFFFFFFFFFFFFFF
 
 typedef unsigned int uint;   NEEDS_CLEANUP
+const uint  max_uint  = (uint)-1;
```
Comment (tschatzl) [inline src/hotspot/share/utilities/globalDefinitions.hpp:442]: The change should use the existing `UINT_MAX` instead of adding a new constant. Instead of a cast of `-1`, one should also use `~0` to get all-ones.

**Review by tschatzl — CHANGES_REQUESTED** [review]
More candidates:

ParGCArrayScanChunk
MarkStackSize(Max) (these should probably be a `size_t`)
VerifyArchivedFields
ScavengeALotInterval
FullGCALotInterval
FullGCALotStart
FullGCALotDummies
VerifyGCLevel
MaxTenuringThreshold
InitialTenuringThreshold
GCDrainStackTargetSize

File: src/hotspot/share/gc/parallel/psAdaptiveSizePolicy.cpp (hunk 1 above)
Comment (afshin-zafari) [reply src/hotspot/share/gc/parallel/psAdaptiveSizePolicy.cpp]: `%u` is used.

File: src/hotspot/share/utilities/globalDefinitions.hpp (hunk 4 above)
Comment (afshin-zafari) [reply src/hotspot/share/utilities/globalDefinitions.hpp]: `UINT_MAX` is used. New definition of max_uint is removed.

**Review by afshin-zafari — COMMENTED** [review]
Suggested changes in Comments are applied including the 'more candidates'.

File: src/hotspot/share/gc/g1/g1Arguments.cpp [hunk 5]
```diff
@@ -229,7 +229,7 @@ void G1Arguments::initialize() {
 
   // By default do not let the target stack size to be more than 1/4 of the entries
   if (FLAG_IS_DEFAULT(GCDrainStackTargetSize)) {
-    FLAG_SET_ERGO(GCDrainStackTargetSize, MIN2(GCDrainStackTargetSize, (uintx)TASKQUEUE_SIZE / 4));
+    FLAG_SET_ERGO(GCDrainStackTargetSize, MIN2(GCDrainStackTargetSize, (uint)TASKQUEUE_SIZE / 4));
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1Arguments.cpp:232]: TASKQUEUE_SIZE should be a constant of type size_t rather than a macro.  GCDrainStackTargetSize should be of type size_t.  Then we don't need any casts here.  But some uses may need cleaning up, either to remove casts or to adjust surrounding types.  I suggest this be split out as a separate RFE to avoid cluttering the simplicity of this PR.

File: src/hotspot/share/gc/parallel/psScavenge.cpp [hunk 6]
```diff
@@ -553,7 +553,7 @@ bool PSScavenge::invoke_no_policy() {
                                                            _tenuring_threshold,
                                                            survivor_limit);
 
-       log_debug(gc, age)("Desired survivor size " SIZE_FORMAT " bytes, new threshold %u (max threshold " UINTX_FORMAT ")",
+       log_debug(gc, age)("Desired survivor size " SIZE_FORMAT " bytes, new threshold %u (max threshold %u)",
```
Comment (kimbarrett) [inline src/hotspot/share/gc/parallel/psScavenge.cpp:556]: Consider also, instead of SIZE_FORMAT we can use "%zu".

File: src/hotspot/share/gc/shared/ageTable.cpp [hunk 7]
```diff
@@ -99,7 +99,7 @@ uint AgeTable::compute_tenuring_threshold(size_t desired_survivor_size) {
   }
 
 
-  log_debug(gc, age)("Desired survivor size " SIZE_FORMAT " bytes, new threshold " UINTX_FORMAT " (max threshold " UINTX_FORMAT ")",
+  log_debug(gc, age)("Desired survivor size " SIZE_FORMAT " bytes, new threshold " UINTX_FORMAT " (max threshold %u)",
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/ageTable.cpp:102]: Consider also, instead of SIZE_FORMAT we can use "%zu".

File: src/hotspot/share/gc/g1/g1Arguments.cpp (hunk 5 above)
Comment (afshin-zafari) [reply src/hotspot/share/gc/g1/g1Arguments.cpp:232]: The change is applied and some cleanup made. Maybe no need to a new RFE, if you agree.

File: src/hotspot/share/gc/parallel/jvmFlagConstraintsParallel.cpp [hunk 8]
```diff
@@ -47,7 +47,7 @@ JVMFlag::Error InitialTenuringThresholdConstraintFuncParallel(uintx value, bool
   if (UseParallelGC && (value > MaxTenuringThreshold)) {
       JVMFlag::printError(verbose,
                           "InitialTenuringThreshold (" UINTX_FORMAT ") must be "
-                          "less than or equal to MaxTenuringThreshold (" UINTX_FORMAT ")\n",
+                          "less than or equal to MaxTenuringThreshold (%u)\n",
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply src/hotspot/share/gc/parallel/jvmFlagConstraintsParallel.cpp:50]: Fixed.

File: src/hotspot/share/gc/parallel/psScavenge.cpp (hunk 6 above)
Comment (afshin-zafari) [reply src/hotspot/share/gc/parallel/psScavenge.cpp:556]: `%zu` is used instead.

File: src/hotspot/share/gc/shared/ageTable.cpp (hunk 7 above)
Comment (afshin-zafari) [reply src/hotspot/share/gc/shared/ageTable.cpp:102]: `%zu` is used.

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 9]
```diff
@@ -431,6 +431,7 @@ const intx  min_intx  = (intx)1 << (sizeof(intx)*BitsPerByte-1);
 const intx  max_intx  = (uintx)min_intx - 1;
 const uintx max_uintx = (uintx)-1;
 
+
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply src/hotspot/share/utilities/globalDefinitions.hpp:434]: Fixed.

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 10]
```diff
@@ -439,7 +440,6 @@ const uintx max_uintx = (uintx)-1;
 
 typedef unsigned int uint;   NEEDS_CLEANUP
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply src/hotspot/share/utilities/globalDefinitions.hpp:441]: Maybe, it is because the `uint` is already defined within the guarded lines 149-154: 
```C++
#ifdef __USE_MISC
/* Old compatibility names for C types.  */
typedef unsigned long int ulong;
typedef unsigned short int ushort;
typedef unsigned int uint;
#endif
```

**Review by coleenp — COMMENTED** [review]
This looks like a good change to me to avoid Wconversion warnings for mis-sized argument values but I'll leave it up to the GC experts.

File: src/hotspot/share/gc/shared/gc_globals.hpp [hunk 11]
```diff
@@ -212,12 +212,12 @@
   /* where does the range max value of (max_jint - 1) come from? */         \
   product(size_t, MarkStackSizeMax, NOT_LP64(4*M) LP64_ONLY(512*M),         \
           "Maximum size of marking stack")                                  \
-          range(1, (max_jint - 1))                                          \
+          range(1, (SIZE_MAX - 1))                                          \
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/gc_globals.hpp:215]: This is a large increase in the max size.  Not at all sure it's appropriate.  Similarly for MarkStackSize below.

File: src/hotspot/share/gc/shared/gc_globals.hpp [hunk 12]
```diff
@@ -677,15 +677,15 @@
...
                                                                             \
   develop(uintx, GCExpandToAllocateDelayMillis, 0,                          \
           "Delay between expansion and allocation (in milliseconds)")       \
                                                                             \
-  product(uintx, GCDrainStackTargetSize, 64,                                \
+  product(size_t, GCDrainStackTargetSize, 64,                               \
           "Number of entries we will try to leave on the stack "            \
           "during parallel gc")                                             \
-          range(0, max_juint)                                               \
+          range(0, SIZE_MAX)                                                \
```
Comment (kimbarrett) [inline src/hotspot/share/gc/shared/gc_globals.hpp:688]: This is a large increase in the max on 64bit platforms.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Comment by kimbarrett** [comment]
> > There are a lot of uses of the options that have casts, many of which ought to be removable with this change. Some of those might pull more tangled strings, but it would be nice if there was a plan to look at those too.
> 
> During this change, I changed all the cases that I found them to be changed. Would you please, give some examples or some 'search criteria' that I can use to find the cases. Then, I will do them here in this PR or create a new RFE for them. 

I spotted some when looking earlier, but haven't re-found any of them.

File: src/hotspot/share/gc/shared/gc_globals.hpp [hunk 13]
```diff
@@ -382,52 +382,52 @@
...
-  product(uintx, AdaptiveTimeWeight,       25,                              \
+  product(uint, AdaptiveTimeWeight,       25,                               \
           "Weight given to time in adaptive policy, between 0 and 100")     \
           range(0, 100)                                                     \
                                                                             \
-  product(uintx, PausePadding, 1,                                           \
+  product(uint, PausePadding, 1,                                            \
           "How much buffer to keep for pause time")                         \
-          range(0, max_juint)                                               \
+          range(0, UINT_MAX)                                                \
```
Comment (coleenp) [inline src/hotspot/share/gc/shared/gc_globals.hpp:395]: Why did you change from max_juint -> UINT_MAX ?  max_juint should have compile without integer size mismatch.

File: src/hotspot/share/gc/shared/gc_globals.hpp (hunk 12 above)
Comment (coleenp) [reply src/hotspot/share/gc/shared/gc_globals.hpp:688]: Yes, you should revert the change from max_juint to SIZE_MAX.  Is the change to size_t a cleanup to remove a cast?

File: src/hotspot/share/utilities/globalDefinitions.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/utilities/globalDefinitions.hpp:440]: I was curious and found it in /usr/include/x86_64-linux-gnu/sys/types.h on ubuntu. I guess it needs cleanup because it might not be defined with USE_MISC?  There are too many int type names to choose from for int.  Why does it say "Old compatibility names for C types."? Should we even be using uint ? I don't see it in https://en.cppreference.com/w/cpp/language/types - should we prefer "unsigned"?

File: src/hotspot/share/gc/shared/gc_globals.hpp (hunk 13 above)
Comment (coleenp) [reply src/hotspot/share/gc/shared/gc_globals.hpp:395]: Or, not in this change, but should we prefer UINT_MAX (found in /usr/include/limits.h) to our globalDefinitions.hpp definitions?

**Review by coleenp — COMMENTED** [review]
I have a few comments.  Narrowing the types of these arguments was to prevent Wconversion warnings where they're used and then later narrowed to 32 bit ints.  From my sampling of some of the arguments changed, there were no casts, which was the problem.  Some of these (sample GCTimeLimit) are promoted back up to uintx but having the field be the narrowest representation allows the code to use it without an implicit conversion to a narrower int type.  So I think this is a good step for cleaning up the int types.

File: src/hotspot/share/gc/shared/gc_globals.hpp (hunk 13 above)
Comment (coleenp) [reply src/hotspot/share/gc/shared/gc_globals.hpp:395]: Now I see Thomas's comments about this.  I guess changing a couple at a time would be ok.

File: src/hotspot/share/gc/shared/gcOverheadChecker.cpp [hunk 14]
```diff
@@ -91,11 +91,11 @@ void GCOverheadChecker::check_gc_overhead_limit(GCOverheadTester* time_overhead,
 
   if (UseGCOverheadLimit) {
     if (gc_overhead_limit_exceeded()) {
-      log_trace(gc, ergo)("GC is exceeding overhead limit of " UINTX_FORMAT "%%", GCTimeLimit);
+      log_trace(gc, ergo)("GC is exceeding overhead limit of %u%%", GCTimeLimit);
       reset_gc_overhead_limit_count();
     } else if (print_gc_overhead_limit_would_be_exceeded) {
       assert(_gc_overhead_limit_count > 0, "Should not be printing");
-      log_trace(gc, ergo)("GC would exceed overhead limit of " UINTX_FORMAT "%% %d consecutive time(s)",
+      log_trace(gc, ergo)("GC would exceed overhead limit of %u %% %d consecutive time(s)",
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply src/hotspot/share/gc/shared/gcOverheadChecker.cpp:98]: Fixed.

File: src/hotspot/share/gc/shared/gc_globals.hpp (hunk 12 above)
Comment (afshin-zafari) [reply src/hotspot/share/gc/shared/gc_globals.hpp:688]: Reverted back to original.

File: src/hotspot/share/utilities/globalDefinitions.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/utilities/globalDefinitions.hpp:440]: This NEEDS_CLEANUP block shouldn't be part of this change.

**Review by tschatzl — APPROVED** [review]
Lgtm, it's a step forward to cleaning up the gc_globals.hpp types/ranges.

**Review by coleenp — APPROVED** [review]
This looks good and definitely a step forward.

**Comment by afshin-zafari** [comment]
Thank you Coleen, Kim, Thomas and Amit for your comments and reviewing this PR.
/integrate
